ion-schema/src/isl/version_based_isl/mod.rs

```diff
@@ -0,0 +1,480 @@
+    user_reserved_fields: UserReservedFields,
+    _version: PhantomData<V>,
+}
+
+impl<V: IslVersionTrait> IslSchemaHeader<V> {
+    pub fn open_content(&self) -> &[Element] {
+        &self.open_content
+    }
+
+    pub fn imports(&self) -> &[IslImport] {
+        &self.imports
+    }
+}
+
+impl IslSchemaHeader<IslV1_0> {
+    pub fn new(open_content: Vec<Element>, imports: Vec<IslImport>) -> Self {
+        Self {
+            open_content,
+            imports,
+            user_reserved_fields: UserReservedFields::default(),
+            _version: Default::default(),
+        }
+    }
+}
+
+impl IslSchemaHeader<IslV2_0> {
+    pub fn new(open_content: Vec<Element>, imports: Vec<IslImport>) -> Self {
+        Self {
+            open_content,
+            imports,
+            user_reserved_fields: UserReservedFields::default(),
+            _version: Default::default(),
+        }
+    }
+
+    pub fn with_user_reserved_fields(
+        self,
+        schema_header_fields: Vec<String>,
+        schema_footer_fields: Vec<String>,
+        type_fields: Vec<String>,
+    ) -> Self {
+        Self {
+            user_reserved_fields: UserReservedFields {
+                schema_header_fields,
+                schema_footer_fields,
+                type_fields,
+            },
+            ..self
+        }
+    }
+
+    pub fn user_reserved_fields(&self) -> &UserReservedFields {
+        &self.user_reserved_fields
+    }
+}
+
+/// Represents a schema footer
+#[derive(Debug, Clone)]
+pub struct IslSchemaFooter {
+    /// Represents open content as `Element`s in the schema footer
+    open_content: Vec<Element>,
+}
+
+impl IslSchemaFooter {
+    pub fn new(open_content: Vec<Element>) -> Self {
+        Self { open_content }
+    }
+}
+
+/// Represents a schema file
+#[derive(Debug, Clone)]
+pub struct IslSchema<V: IslVersionTrait> {
+    /// Represents an id for the given ISL model
+    id: String,
+    /// Represents the schema header if its present which contains imports and any open content provided in the header.
+    /// For ISL 2.0 schema header also contains user reserved fields.
+    /// If the schema header is not present this is None.
+    schema_header: Option<IslSchemaHeader<V>>,
+    /// Represents the schema footer if its present which contains any open content provided in the footer.
+    /// If the schema footer is not present this is None.
+    schema_footer: Option<IslSchemaFooter>,
+    /// Represents all the IslType defined in this schema file.
+    /// For more information: https://amazon-ion.github.io/ion-schema/docs/isl-1-0/spec#type-definitions
+    types: Vec<IslType<V>>,
+    /// Represents all the inline IslImportTypes in this schema file.
+    inline_imported_types: Vec<IslImportType>,
+    /// Represents open content as `Element`s
+    /// Note: This doesn't preserve the information about where the open content is placed in the schema file.
```

Comment:
This is reasonable, but has interesting effects on how we serialize a schema later. We should talk through the impact it'll have. No `TODO` here.